### PR TITLE
Add rough gun code

### PR DESCRIPTION
**Pull request by MagiMaster**
Add server/client code to handle gun aiming and building swapping.

### CODE REVIEW

File: src/client/Gun.client.lua [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+-- The gun model should contain at least:
+--   Gun: Model
+--     Handle: Part
+--       Attachment: Attachment
+--       WeldConstraint: WeldConstraint
+--     Target: Part
+--       Attachment: Attachment
+--       Beam: Beam (connected to the two attachments above)
+--       ParticleEmitter: ParticleEmitter
+localPlayer.CharacterAdded:Connect(function(character)
```
Comment (ChipioIndustries) [inline src/client/Gun.client.lua]: Try to avoid using anonymous functions like this.

File: src/client/Gun.client.lua [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+	local firing = false
+
+	local function updateTarget()
+		local mousePos = UserInputService:GetMouseLocation()
+		local viewRay = workspace.CurrentCamera:ViewportPointToRay(mousePos.X, mousePos.Y)
+		local result = workspace:Raycast(viewRay.Origin, viewRay.Direction * 1000)
+		return result and result.Position
+	end
+
+	UserInputService.InputBegan:Connect(function(input)
```
Comment (ChipioIndustries) [inline src/client/Gun.client.lua]: This connection is being made every time the character respawns - it should be moved outside of the CharacterAdded callback, or the connection handle should be saved and then disconnected when the player dies.

File: src/client/Gun.client.lua [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+				local pos = updateTarget()
+				if pos then
+					convert:FireServer(true, pos)
+				end
+				wait(0.1)
+			end
+		end
+	end)
+
+	UserInputService.InputEnded:Connect(function(input)
```
Comment (ChipioIndustries) [inline src/client/Gun.client.lua]: Same comment as above.

File: src/client/Gun.client.lua [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+		if input.UserInputType == Enum.UserInputType.MouseButton1 then
+			convert:FireServer(true)
+
+			firing = true
+			while firing do
+				local pos = updateTarget()
+				if pos then
+					convert:FireServer(true, pos)
+				end
+				wait(0.1)
```
Comment (ChipioIndustries) [inline src/client/Gun.client.lua]: wait() has been deprecated in favor of task.wait(). They function identically but task.wait() uses the more reliable task scheduler

File: src/server/Gun.server.lua [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+				replaceBuilding(obj, obj.NewVersion.Value)
+				stability = 50
+			end
+		end
+		obj:SetAttribute("Stability", stability)
+	end
+end
+
+convert.OnServerEvent:Connect(function(player, start, pos)
+    local gun = player.Character.Gun
```
Comment (ChipioIndustries) [inline src/server/Gun.server.lua]: This line is indented with spaces instead of tabs

File: src/server/Gun.server.lua [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+		TweenService:Create(target, info, {Position = pos}):Play()
+	end
+end)
+
+for _, part in workspace:GetChildren() do
+	if part:IsA("Model") and part:FindFirstChild("OldVersion") then
+        -- Temporary. This should randomize buildings.
+        part:SetAttribute("Stability", 100)
+		replaceBuilding(part, part.NewVersion.Value)
+	end
```
Comment (ChipioIndustries) [inline src/server/Gun.server.lua]: Some of these lines are also indented with spaces

File: src/server/Gun.server.lua [hunk 7]
```diff
@@ -0,0 +1,84 @@
+local Players = game:GetService("Players")
+local ReplicatedStorage = game:GetService("ReplicatedStorage")
+local Teams = game:GetService("Teams")
+local TweenService = game:GetService("TweenService")
+
+local convert = ReplicatedStorage:WaitForChild("Convert")
```
Comment (ChipioIndustries) [inline src/server/Gun.server.lua]: WaitForChild() shouldn't be necessary on the server since the contents are already stored there.

**Review by ChipioIndustries — CHANGES_REQUESTED** [review]
A few issues/requests:

I noticed when testing that if I shoot far away from myself, or in a few other unclear circumstances, the beam from my gun will shoot directly towards the world origin (0,0,0) no matter where I'm pointing. My best guess is that this happens whenever something in the updateTarget function returns nil.

The code is styled and structured pretty well but I would like it to be assembled from modulescripts instead to allow for more interoperability. You can create server/local scripts that then require the modules on startup.

Good stuff overall!

File: src/client/Gun.client.lua (hunk 2 above)
Comment (MagiMaster) [reply src/client/Gun.client.lua]: Added some cleanup logic to all the connections.

File: src/server/Gun.server.lua (hunk 5 above)
Comment (MagiMaster) [reply src/server/Gun.server.lua]: Oops.

**Review by ChipioIndustries — COMMENTED** [review]
I think I was unclear in how I described a single-script architecture - I didn't mean that the game code should all go in a single script with the exception of classes, I meant that all game code should go into modules and the only job of the single script is to start up those modules. You can see an example of this in this project: https://github.com/ChipioIndustries/dragon-game/blob/master/src/server/ServerStart.server.lua

**Pull request merged**